### PR TITLE
Add rawtcp plugin to fluentd

### PR DESCRIPTION
**Pull request by antoineHC**
Hello, 

In order to send logs to external logstash service from ```user/output-extra-*.conf```, I added https://github.com/uken/fluent-plugin-out_rawtcp fluentd plugin. 
This plugin implements a simple TCP forward, without heartbeats or any other out-of-band checks. 

Regards,

### CODE REVIEW

**Comment by richm** [comment]
Where is this source code from?  What is the license?  We also require a test to demonstrate this functionality.  There is no fluentd config for this so a user would have to know how to hack and configure the fluent.conf configmap to use this.  Finally, it looks as though this is using msgpack to encode the payload - does logstash understand msgpack?

**Comment by antoineHC** [comment]
Hello, 

@richm I added the source and licence in the commit https://github.com/openshift/origin-aggregated-logging/pull/968/commits/b670d5c4d421a408e559ccadeaecf7ecf85d939e

In order to make it works you can use the default config map from ansible playbook and put the following code in `configs.d/user/output-extra-1.conf`

```
<store>
    type rawtcp 
    flush_interval 1
    <server>
      name logstash
      host my-logstash.external.com
      port 9400
    </server>
</store>
```
The msgpack must be used in logstash using the built in [fluent codec](https://www.elastic.co/guide/en/logstash/current/plugins-codecs-fluent.html) 
Logstash example configuration: 
```
input {
    tcp {
        port => 9400
        codec => fluent
    }
}

output {
    elasticsearch {
        hosts => ["localhost:9200"]
        index => "logstash-%{+YYYY.MM.dd}"
        document_type => "logs"
    }
}
```
@jcantrill About the secure communication to the destination it is the same as syslog protocol 

Regards,

**Comment by richm** [comment]
Why would we consume the source code rather than the rubygem?  https://rubygems.org/gems/fluent-plugin-out_rawtcp

**Comment by antoineHC** [comment]
The gem install doesn't work because in the gemspec, the latest version of fluentd is referenced and it depends on Ruby 2.1 which is not a Ruby version available in the RHEL Fluentd image. 
https://github.com/uken/fluent-plugin-out_rawtcp/blob/master/fluent-plugin-out_rawtcp.gemspec#L18 
As we rely on the Fluentd install in the container, the plug-in works and the only way to install it is to copy directly the sources in the image.

**Comment by richm** [comment]
@jcantrill @portante This gives us the ability to forward logs to logstash.  I vote for including this.  Approve?

**Comment by antoineHC** [comment]
Ping @jcaianirh @portante any news on the subject ?

**Comment by jcantrill** [comment]
Is there a test we can add for this plugin?  My concern is we merge it, someone uses it, we have no test that validates it functionally works as advertised and then we are on the hook to support it.

**Comment by antoineHC** [comment]
@openshift-bot /retest

File: test/out_rawtcp.sh [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+  fi
+
+  $mycmd raw-tcp test finished at $( date )
+
+  # Clean up only if it's still around
+  os::log::debug "$( oc delete service/logstash 2>&1 || : )"
+  os::log::debug "$( oc delete deploymentconfig/logstash 2>&1 || : )"
+  
+  os::cmd::expect_success flush_fluentd_pos_files
+  os::log::debug "$( oc label node --all logging-infra-fluentd=true 2>&1 || : )"
```
Comment (richm) [inline test/out_rawtcp.sh]: need to wait for fluentd to be running here
```
os::cmd::try_until_text "oc get pods -l component=fluentd" "^logging-fluentd-.* Running "
```

File: test/out_rawtcp.sh [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+os::log::info Starting raw-tcp test at $( date )
+
+# make sure fluentd is working normally
+os::cmd::try_until_text "oc get pods -l component=fluentd" "^logging-fluentd-.* Running "
+fpod=$( get_running_pod fluentd )
+os::cmd::expect_success wait_for_fluentd_to_catch_up
+
+create_forwarding_logstash
+update_current_fluentd
+cleanup
```
Comment (richm) [inline test/out_rawtcp.sh]: remove this call to `cleanup` - the `trap EXIT` will call this upon exit from the script, no matter how that exit happens

File: test/out_rawtcp.sh (hunk 1 above)
Comment (antoineHC) [reply test/out_rawtcp.sh]: Added the command, thanks

**Comment by antoineHC** [comment]
@richm Do I need to get the tests green ? They seem to run an incomplete test set when I launch them manually :(

**Comment by richm** [comment]
@antoineHC looks like the CI system is currently undergoing maintenance - should be fixed soon

**Comment by stevekuznetsov** [comment]
v

**Comment by richm** [comment]
you keep settin'm up, I'll keep settin'm down . . .
/retest

**Comment by antoineHC** [comment]
One last time 
/retest

**Comment by richm** [comment]
Whatever the failure is here, it doesn't appear to be a flake - it is persistent, always in the same place in the utf8 test.  I have no idea why.

**Comment by richm** [comment]
@antoineHC looks good - please rebase to the latest master and squash your commits into a single commit

**Comment by richm** [comment]
lots of fluentd failures
/retest

**Comment by antoineHC** [comment]
Oh yeah all green 🎉 🍾

**Comment by richm** [comment]
@jcantrill @portante ok?  lgtm

**Comment by richm** [comment]
@portante do you still require changes?

**Comment by antoineHC** [comment]
@richm Is there any change in the namespace name ? Before having the lgtm label, the tests were green with `namespace: logging` in my logstash deployment. 

Since then, I get an error message 
> namespaces "logging" not found

**Comment by richm** [comment]
yes - the default logging namespace is now `openshift-logging`

**Comment by richm** [comment]
strange curator flakes . . .
/retest

**Pull request merged**